### ISSUE TITLE
Can openssl lib support Subject Key Attestation Evidence Extension?

### ISSUE THREAD

**Issue by stardrift1** [openssl/openssl#12515]
@mattcaswell 
Now i want to request a certificate to prove that my private key is stored in the security chip(such as TPM),and Meets the requirements of TCG（it's import to our Product ）,SO I should make a certificate message which carry the Subject Key Attestation Evidence Extension which oid is joint-iso-itu-t(2) international-organizations(23) tcg(133)ce(6) skae(1) subjectKeyAttestationEvidence(1)，but now i can't find it's support in openssl lib.
so,gentle man ,pls help check if it should add to openssl lib as a new stable feature or it just should Implemented at the application layer ？

**Comment by beldmit** [comment]
If you just want to add a specific extension value to X509 certificate, it's worth looking to man x509v3_config

**Comment by stardrift1** [comment]
> If you just want to add a specific extension value to X509 certificate, it's worth looking to man x509v3_config

I Want to do CMP req with the crmf messgge carry the extension,it can be done with  x509v3_config?

**Comment by mattcaswell** [comment]
@DDvO

**Comment by DDvO** [comment]
I've added an example of using the CMP `-reqexts` option in PR #12534.
This shows how to conveniently add X.509 extensions to CMP requests.
The syntax in the config file is the same as for the `openssl req` app.

@stardrift1, this way you should be able to specify also the SKAE extension.

**Comment by stardrift1** [comment]
> The syntax in the config file is the same as for the `openssl req` app.

I have get how to use config file to add extension ,but i want to use openssl lib to creat my own application which can use some function to creat the SKAE inner structure ,and use some other funtion to encode the SKAE inner structure to der Format and then call OSSL_CMP_CTX_set0_reqExtensions() function to add the der Format ext to ctx,then add to cmp message.
maybe can X509V3_EXT_METHOD for SKAE ext,and use X509V3_EXT_add（）function to add this method to ext_list,then i can call the function to do the SKAE ext new/i2d/d2i/free.
I think this is import to TCG so i make this issure to consult if this feature can be a stable part of the openssl lib,but not the app layer impl.

**Comment by DDvO** [comment]
@shlomif, good that you now know how to add an SKAE extension using a config file, and hopefully you can use that as a model on how to do this via the CMP and X.509 API.
Yes, the `OSSL_CMP_CTX_set0_reqExtensions()` function is the right CMP API function to add X.509 extensions to a CMP request.
I'm not an expert on constructing specific extensions, but I suspect that maybe this can be done in a simpler way than you mentioned above, namely by using the `X509V3_EXT_i2d()`, `X509v3_add_ext()`, and `X509_EXTENSION_free()` functions.

**Comment by shlomif** [comment]
@DDvO : you probably meant @stardrift1 - not me.

**Comment by DDvO** [comment]
> @DDvO : you probably meant @stardrift1 - not me.

Opps, yes, I meant @stardrift1

**Comment by stardrift1** [comment]
> I'm not an expert on constructing specific extensions,

Is it necessary to add this feature to the ssl lib,or  it just should be Implemented at the application layer?
Would you konw who can decide this?

**Comment by DDvO** [comment]
I'm not sure if the SKAE extension is so fundamental that it would be worth adding specific support for it in OpenSSL - what do other @openssl/committers think?
If so, it should be done not in the SSL lib but in the X.509 portion of the crypto lib, and @stardrift1 you could provide a pull request with a suggested implementation.

**Comment by beldmit** [comment]
If it requires non-trivial printing or creating, providing it via X.509 library is near to the only way to deal with it comfortably.

**Comment by stardrift1** [comment]
> non-trivial printing or creating

I think it requires ,i even need creat some function to add TPMCertifyInfo to the SKAE ext,Which need compute the signature first.If Implemented at the application layer by our user ,it a little Complex and maybe lead Redundant work.
TPMCertifyInfo ::= SEQUENCE {
 CertifyInfo BIT STRING,
-- TPM_CERTIFY_INFO structure as a bit string
 signature BIT STRING }

**Comment by DDvO** [comment]
@stardrift1, can you provide example code, or even better a pull request implementing this?

**Comment by stardrift1** [comment]
Maybe later@DDvO

**Comment by stardrift1** [comment]
> @stardrift1, can you provide example code, or even better a pull request implementing this?

i want to add this ext method to standard standard_exts,it can be done?

**Comment by DDvO** [comment]
I think it would be best to the ` standard_exts` in libcrypto, not just to an app.
Then we can directly review and take over the support for the new extension.

**Comment by nhorman** [comment]
Marking as inactive to be closed at the end of 3.4 dev, barring further input

**Comment by DDvO** [comment]
As key attestation in CSRs has gained quite some significance recently, see https://datatracker.ietf.org/doc/draft-ietf-lamps-csr-attestation/
 we should not close this, on the contrary more actively pursue it

**Comment by nhorman** [comment]
Reading over this issue, heres the history that I take from it:

[@stardrift1 wanted to add the SKAE extention to a certificate programatically](https://github.com/openssl/openssl/issues/12515#issue-664368277)

[The question of how to do that was answered](https://github.com/openssl/openssl/issues/12515#issuecomment-663847321)

[@stardrift1 clarified what they want to do](https://github.com/openssl/openssl/issues/12515#issuecomment-663851175)

[you asked for example code, which, for whatever reason wasn't, or couldn't be provided](https://github.com/openssl/openssl/issues/12515#issuecomment-668423796)

[@stardrift1 offered to implement support for the extension, which it appears, for any nunmber of reasons, never matrialized](https://github.com/openssl/openssl/issues/12515#issuecomment-669649572)

The issue then sat unaddressed for 4 years.


You're statement of "we should more actively pursue this" isn't valid, it certainly is @DDvO, but I'm looking at an issue here in which the reporter offered to create a patch for this, and then for whatever reason (lack of time/bandwith, etc), didn't get to it.  Nor can I find any other issue or pull request referencing SKAE extentions.  

So in the end, while it would be nice to get to everything requested, I'm hard pressed to see this as a priority.

If there is a community volunteer willing to take this on again, thats a different story, and I'm happy to adjust this accordingly.

**Comment by DDvO** [comment]
Thanks for your further thoughts.
Right, the "we" that I wrote was not to the point (at least was too ambiguous), should more generally have read: "someone".

I'm pretty sure that demand for key attestation support in CSRs will come up rather soonish,
even if the original proposing person @stardrift1 lost interest or does not have the capability to do it.
In case this FR got closed meanwhile, it could be re-opened (at least upon request).

**Comment by nhorman** [comment]
@DDvO agreed.  I would expect to see this reopened or a new issue created, should the need for SKAE extensions become a higher priority, at which point we will certainly re-evaluate for higher priority development and inclusion
